### PR TITLE
bdn module

### PR DESCRIPTION
**Pull request by YingxiaoKong**
This is an update of the bdn module which is used to make predictions of data with model uncertainty.

### CODE REVIEW

File: paraatm/bdn/_bdn.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        """
+        if flag_2 =='rnn':
+            model = self._get_rnn_model()
+        if flag_2 =='dnn':
+            model = self._get_dnn_model()
+        if flag_2 =='custom':
+            model = custom_model
+        #print(model.summary())
+        model.fit(self.x_train,self.y_train,self.batch_size,
+                      self.epochs,verbose = False)
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:199]: Would it be better to move the `model.fit` call under `__init__` so that fitting is not repeated every time `pred` is called?  Depending on what makes the most sense, you could either take the model type flag as an input to `__init__`, or you could fit all model types in `__init__` and then set `model` to point to the appropriate structure in `pred`.

File: paraatm/bdn/_bdn.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            By default: it's set as 'True'
+            flag_1 == True: prediction with uncertainty
+            flag_1 == False: deterministic result
+        
+        flag_2: str
+            By default, it's set as 'rnn'
+            flag_2 == 'rnn': generate results for the embedded rnn model
+            flag_2 == 'dnn': generate results for the embedded dnn model
+            flag_2 == 'custom': generate results for the customized model
+            When flag_2 == 'custom', a customized model must be provided
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:180]: Can you use more descriptive names for these?  Instead of `flag_1` and `flag_2`, perhaps `get_stdev` (or `get_var`) and `model_type`?

File: paraatm/bdn/_bdn.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        
+        custom_model: None is set by default, otherwise is a  customized sequential model
+        
+        Returns
+        -------
+        An array which is predictions for the test data
+        The array has a shape as num_test_data by num_iter
+        num_test_data:  Number of test data
+        num_iter: number of iterations
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:189]: Is the return type different when `flag_1` for is set for prediction with uncertainty?  If so, please document how the uncertainty is represented in the return type.

File: paraatm/bdn/_bdn.py (hunk 1 above)
Comment (YingxiaoKong) [reply paraatm/bdn/_bdn.py:202]: Hi John,

Thank you for the comments and suggestion.

The model can only be fitted in after the _get_dnn_model or _get_rnn_model is called and returned an model. The neural network model I'm using right now is a sequential model, which means that it starts with input layer, then hidden layers, then output layer. After the model structure is built, it would be fit in. As the DNN model and RNN model doesn't use same data. In DNN, the input data is 2D; in RNN, the input data is 3D. So probably there is no redundancy here.  For 'you could fit all model types in __init__ and then set model to point to the appropriate structure in pred', I don't really understand it very well. Could you give me a more detailed explanation or a simple example?

File: paraatm/bdn/_bdn.py (hunk 2 above)
Comment (YingxiaoKong) [reply paraatm/bdn/_bdn.py]: Hi John,

Thanks for the comments. ' flag_1' now is changed to 'pred_type'. 'flag_2' is changed to 'model_type'.

File: paraatm/bdn/_bdn.py (hunk 3 above)
Comment (YingxiaoKong) [reply paraatm/bdn/_bdn.py:192]: Hi John,

A detailed explanation for how uncertainty is included is added to the document string.

File: paraatm/bdn/_bdn.py (hunk 2 above)
Comment (jmcfarland-swri) [reply paraatm/bdn/_bdn.py]: Two comments:
* Why did you move `pred_type` to be an argument to `__init__` instead of `pred`?  I think it makes more sense to be an argument of `pred`.  That way, the user has the flexibility to get different types of predictions after creating the model.
* The variable name is not consistent with its type.  `True` and `False` are not types of predictions.  It would be more logical to name the variable in a way that is consistent with the `True` and `False` values.  That is why my original suggestion was `get_stdev` or `get_var` (get standard deviation or variance).  This is more intuitive because it is clear that `get_stdev=True` means "yes, I want to also retrieve the standard deviation for the prediction".  It is not intuitive why `pred_type=True` would mean that the result should include uncertainty.  Other options would be `return_stdev`, `return_var`, `return_uncertainty`, etc.

File: paraatm/bdn/_bdn.py (hunk 3 above)
Comment (jmcfarland-swri) [reply paraatm/bdn/_bdn.py:192]: This doesn't actually address my original comment.  I cannot tell from reading the documentation how uncertainty is represented in the return type.  The question is not about the method used to calculate uncertainty, it is about the type of data that is returned from the `pred` function when the user has requested uncertainty via the flag (originally `flag_1`).

For example, is the return value a list that includes both a mean and a standard deviation?  Is it just a standard deviation?  Is it a variance?  Is it an array of means and standard deviations?  Is the return type different when different values are provided to `flag_1`?

Also, this description of the return type belongs in the `Returns` section of the docstring.  For an example, refer to the documentation for `SklearnGPRegression.__call__` in [gp.py](../blob/master/paraatm/rsm/gp.py).

File: paraatm/bdn/_bdn.py (hunk 1 above)
Comment (jmcfarland-swri) [reply paraatm/bdn/_bdn.py:202]: Here is the basic idea of how this could be done:

```python
class Bdn:
    def __init__(self):
        # Fit models so that they are ready to use when predictions
        # are requested
        self.rnn = self._get_rnn_model()
        self.dnn = self._get_dnn_model()
    def pred(self, x, model_type='rnn'):
        if model_type == 'rnn':
            return self.rnn.predict(x)
        elif model_type == 'dnn':
            return self.dnn.predict(x)
```

If you were concerned that you might not need both DNN and RNN to be fit, you could do something like this:
```python
class Bdn:
    def __init__(self):
        # Don't fit models until they are first needed
        self.rnn = None
        self.dnn = None
    def pred(self, x, model_type='rnn'):
        if model_type == 'rnn':
            if self.rnn is None:
                # Fit the model and store it so that it is available
                # the next time it is needed
                self.rnn = self._get_rnn_model()
            return self.rnn.predict(x)
        elif model_type == 'dnn':
            if self.dnn is None:
                self.dnn = self._get_dnn_model()
            return self.dnn.predict(x)
```

Again, I don't know the in's and out's of this particular method that you are using.  However, this is normally how data driven models would be implemented.  Fitting can be expensive, so you want to only do it once.  For example, what if the user wants to do something like this:
```python
model = Bdn(Xtrain, Ytrain)
results = model.predict(x)
# Do some calculations, then make another prediction
results = model.predict(x)
# Do some calculations, then comptue a series of predictions
for i in range(1000):
    results = model.predict(...)
```

Your code would require re-fitting the model every time the user calls `pred`, but that is unecessary.  It should be possible to fit the model once and then re-use it for predictions as many times as the user wants in the same program.

File: paraatm/bdn/_bdn.py (hunk 3 above)
Comment (YingxiaoKong) [reply paraatm/bdn/_bdn.py]: Hi John,

I just updated the module. A more detailed description of how uncertainty should be obtained is added to the document. 
pred_type has been renamed as get_var.
model.fit method has been moved to ___init__  so that the model only need to be fitted once for the deterministic and uncertainty case.
x_test and y_test parameter have been moved to 'pred' method, so that the user can reuse the model for different test data  at 'pred' method.

File: paraatm/bdn/_bdn.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+            model.add(Dropout(self.odrop))
+        model.add(Dense(1,activation='linear',
+                        kernel_regularizer=l2(self.weight_decay),
+                        bias_regularizer=l2(self.weight_decay)))
+        optimizer_=Adam(self.lr)
+        model.compile(loss='mse',metrics=['mse'],optimizer=optimizer_)
+        return model        
+        
+
+    def pred(self, x_test, y_test, get_var = True,iter_=5):
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:181]: `y_test` is not used in this function, right?  If so, please remove it as an argument.

File: paraatm/bdn/_bdn.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            Number of realizations for each sample
+            By default, it's set as 5
+            This is only used when get_var == True
+        
+        Returns
+        -------
+        An array which is prediction for the test data
+        The array has a shape as num_iter by num_test_data
+        num_test_data: number of test data
+        num_iter: number of iterations
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:213]: I still do not think this is a complete description of the return value.

1. When I call this function with `get_var=False`, the shape of the results is not as described in these comments.
2. The first line states that the return value is "an array which is prediction for the test data".  This applies only to the case when `get_var=False`, right?  If `get_var=True`, do I get variance values instead of predicted values (means)?

File: paraatm/bdn/_bdn.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+        """
+        model = self.model
+        if get_var==True:
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:217]: Conventional usage is simply to check `if get_var:` (not to compare against `True`)

File: paraatm/bdn/_bdn.py (hunk 4 above)
Comment (YingxiaoKong) [reply paraatm/bdn/_bdn.py]: Hi John,

It's not used for the pred method. I just removed it.

File: paraatm/bdn/_bdn.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            f = K.function([model.layers[0].input,
+                                K.learning_phase()],[model.layers[-1].output])
+            results = []
+            for i in range(iter_):
+                results.append(np.squeeze(f([x_test,1])))
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:222]: Optional simplification, but typically we try and avoid `for` loops in Python.  This could be replaced by a list comprehension:
```suggestion
            results = np.array([np.squeeze(f([x_test,1])) for i in range(iter_)])
```

File: paraatm/tests/testBdn.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                
+    def test_RNN(self):
+        rnn_model = Bdn(RNN_x_train,RNN_y_train)
+        rnn_y_test_pred = rnn_model.pred(RNN_x_test, RNN_y_test)
+        self.assertEqual(rnn_y_test_pred.shape[0],5)
+    ### test the DNN network
+    def test_DNN(self):
+        dnn_model = Bdn(DNN_x_train,DNN_y_train,model_type = 'dnn' )
+        dnn_y_test_pred = dnn_model.pred(DNN_x_test, DNN_y_test)
+        self.assertEqual(dnn_y_test_pred.shape[0],5)
```
Comment (jmcfarland-swri) [inline paraatm/tests/testBdn.py]: Two changes that I would suggest:

1. Test `pred` with both `get_var=True` and `get_var=False`
2. Check the size of both dimensions of the result, not just `shape[0]`.  Your prediction function is a little bit unusual in its return value type, so I think it would be good to have some additional testing here to make sure it doesn't break in the future.

File: paraatm/bdn/_bdn.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            model.add(Dropout(self.odrop))
+        model.add(Dense(1,activation='linear',
+                        kernel_regularizer=l2(self.weight_decay),
+                        bias_regularizer=l2(self.weight_decay)))
+        optimizer_=Adam(self.lr)
+        model.compile(loss='mse',metrics=['mse'],optimizer=optimizer_)
+        return model        
+        
+
+    def pred(self, x_test, get_var = True,iter_=5):
```
Comment (jmcfarland-swri) [inline paraatm/bdn/_bdn.py:181]: Is `get_var=True` the best default value?  I'm still a little bit unclear on exactly what this does (I think it returns a variance instead of a prediction).  If I understand this correctly, it seems that returning the prediction (`get_var=False`) would be a more logical default.

**Comment by YingxiaoKong** [comment]
get_var = True returns the prediction with multiple realizations. By setting get_var= True, the model will consider the testing phase as learning phase and dropout is applied during testing phase, in which model uncertainty is included. To get the variance, you can simply use np.var(predictions). I have also considered of only returning the variance and mean value, but in that way the user can’t view the distributions of the raw predictions. Maybe I should return the raw data, mean and variance.

When get_var = False, there is no dropout during testing, dropout only happens during training phase. The model weights are constant after training, so the prediction is deterministic.

Thanks for all your comments. I’m going to address them one by one.

在 2020年6月15日，下午5:42，jmcfarland-swri <notifications@github.com> 写道：

﻿

@jmcfarland-swri commented on this pull request.

________________________________

In paraatm/bdn/_bdn.py<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fymlasu%2Fpara-atm%2Fpull%2F17%23discussion_r440486643&data=02%7C01%7Cyingxiao.kong%40vanderbilt.edu%7Cd3b6e48a386f4e7c7a8008d8117d5bfe%7Cba5a7f39e3be4ab3b45067fa80faecad%7C0%7C1%7C637278577396420615&sdata=P%2FZrW217r6bCPtZxseheowRfH9qIb2UE0GMqsj7paAI%3D&reserved=0>:

> +                       bias_regularizer=l2(self.weight_decay),
+                        kernel_initializer='random_uniform',
+                       ))
+        model.add(Dropout(self.idrop))
+        model.add(Activation('relu'))
+        if self.odrop:
+            model.add(Dropout(self.odrop))
+        model.add(Dense(1,activation='linear',
+                        kernel_regularizer=l2(self.weight_decay),
+                        bias_regularizer=l2(self.weight_decay)))
+        optimizer_=Adam(self.lr)
+        model.compile(loss='mse',metrics=['mse'],optimizer=optimizer_)
+        return model
+
+
+    def pred(self, x_test, get_var = True,iter_=5):


Is get_var=True the best default value? I'm still a little bit unclear on exactly what this does (I think it returns a variance instead of a prediction). If I understand this correctly, it seems that returning the prediction (get_var=False) would be a more logical default.
